modules/dex/libdexc/adapter.go

```diff
@@ -0,0 +1,266 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync/atomic"
+
+	"decred.org/dcrdex/client/core"
+	"decred.org/dcrdex/client/webserver"
+	"decred.org/dcrdex/dex"
+	"github.com/decred/slog"
+
+	_ "decred.org/dcrdex/client/asset/btc" // register btc asset
+	_ "decred.org/dcrdex/client/asset/dcr" // register dcr asset
+)
+
+type callHandler func(json.RawMessage) (string, error)
+
+func reply(thing interface{}) (string, error) {
+	b, err := json.Marshal(thing)
+	if err != nil {
+		return "", err
+	}
+	return string(b), nil
+}
+
+func replyWithErrorCheck(thing interface{}, err error) (string, error) {
+	if err != nil {
+		return "", err
+	}
+	return reply(thing)
+}
+
+// CoreAdapter manages a Core instance and possibly a Server instance for
+// a library user.
+type CoreAdapter struct {
+	ctx           context.Context
+	kill          context.CancelFunc
+	inited        uint32
+	logLevel      slog.Level
+	serverRunning uint32
+	core          *core.Core
+	webServer     *dex.ConnectionMaster
+
+	preInitMethods map[string]callHandler
+	directMethods  map[string]callHandler
+}
+
+func NewCoreAdapter() *CoreAdapter {
+	c := &CoreAdapter{}
+
+	c.preInitMethods = map[string]callHandler{
+		"IsInitialized": c.isInitialized,
+	}
+
+	c.directMethods = map[string]callHandler{
+		// Some control functions.
+		"startServer": c.startServer,
+		"shutdown":    c.shutdown,
+		// Pass-throughs to Core
+		"Init":         c.init,
+		"CreateWallet": c.createWallet,
+		"UpdateWallet": c.updateWallet,
+		"User":         c.user,
+		"Register":     c.register,
+		"Login":        c.login,
+		"Logout":       c.logout,
+		"DexConfig":    c.getDexConfig,
+	}
+
+	return c
+}
+
+func (c *CoreAdapter) startCore(raw json.RawMessage) error {
+	if !atomic.CompareAndSwapUint32(&c.inited, 0, 1) {
+		return fmt.Errorf("already initialized")
+	}
+
+	form := new(struct {
+		DBPath   string      `json:"dbPath"`
+		Net      dex.Network `json:"net"`
+		LogLevel slog.Level  `json:"logLevel"`
+	})
+	if err := json.Unmarshal(raw, form); err != nil {
+		return err
+	}
+
+	err := os.MkdirAll(filepath.Dir(form.DBPath), 0777)
+	if err != nil {
+		return err
+	}
+
+	c.ctx, c.kill = context.WithCancel(context.Background())
+	ccore, err := core.New(&core.Config{
+		DBPath: form.DBPath,
+		Net:    form.Net,
+		Logger: dex.StdOutLogger("CORE", form.LogLevel),
+	})
+	if err != nil {
+		return fmt.Errorf("error creating client core: %v", err)
+	}
+	c.core = ccore
+	c.logLevel = form.LogLevel
+
+	go ccore.Run(c.ctx)
+	<-ccore.Ready()
```

Comment:
Currently `shutdown` will return before Core has actually finished shutting down, which can be bad if the whole Decrediton process and this module die before cleanup.  I'd suggest the following:

```go
	c.wg.Add(1)
	go func() {
		defer c.wg.Done()
		ccore.Run(c.ctx)
	}()
	<-ccore.Ready()
```

and in `shutdown`:

```go
	c.kill()
	c.wg.Wait()
```